waterfall/golang/net/qemu/control.go

```diff
@@ -0,0 +1,347 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package qemu provides implementations of a net.Conn and net.Listener backed by qemu_pipe
+package qemu
+
+import (
+	"bufio"
+	"io"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	control_socket "github.com/google/waterfall/proto/control_socket_go_proto"
+)
+
+// Conn implements the net.Conn interface on top of a qemu_pipe
+type ControlSocketConn struct {
+	// Backed by qemu_pipe
+	wb *bufio.Writer
+	rb *bufio.Reader
+	cl io.Closer
+
+	cb *ControlSocket
+	fd uint32
+
+	// Dummy addr
+	addr net.Addr
+
+	closedReads  bool
+	closedWrites bool
+	closed       bool
+
+	readsCloseChan  chan struct{}
+	writesCloseChan chan struct{}
+
+	readLock  *sync.Mutex
+	writeLock *sync.Mutex
+	closeLock *sync.Mutex
+}
+
+// Read reads from from the Conn connection.
+// Note that each message is prepended with the size, so we need to keep
+// track of how many bytes we need to read across Read calls
+func (q *ControlSocketConn) Read(b []byte) (int, error) {
+	q.readLock.Lock()
+	defer q.readLock.Unlock()
+	return q.rb.Read(b)
+}
+
+// Write writes the contents of b to the underlying connection
+func (q *ControlSocketConn) Write(b []byte) (int, error) {
+	q.writeLock.Lock()
+	defer q.writeLock.Unlock()
+
+	if q.closedWrites {
+		return 0, errClosed
+	}
+
+	n, err := q.wb.Write(b)
+	if err != nil {
+		return n, err
+	}
+	return n, q.wb.Flush()
+}
+
+// Close closes the connection
+func (q *ControlSocketConn) Close() error {
+	q.closeLock.Lock()
+	defer q.closeLock.Unlock()
+
+	if q.closed {
+		return errClosed
+	}
+
+	q.closed = true
+	q.CloseWrite()
```

Comment:
Do the errors returned by any of these close calls matter? Should we just ignore all of these errors or log them?

Reply:
Adding support for proper error handling.